firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,446 @@
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with a complete 2-byte header, and paylaod buffer as '0x00' consistent with the length field of the header is parsed")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with a complete 2-byte header, and a non-empty payload buffer where the length field of the header is inconsistent with the actual length of the payload buffer")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is not equal to the size of the payload buffer in the input_buffer body")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+    WHEN("A body with sequence field of the header inconsistent with sequence given in the constructor of the datagram")
+      THEN("The parse method reports ok status")
+      THEN("After the parse method is called, The value returned by the seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The payload returned from the payload accessor method is equal to the payload from the input_buffer body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the Datagram constructor") 
+      THEN("the payload given in the Datagram constructor is independent of the input buffer body")
+
+
+Scenario: Datagram correctly preserves data in roundtrip writing and parsing
+  GIVEN("A datagram constructed with a non-empty payload buffer and sequence number equal to 0")
```

Comment:
I'd prefer initializing with a non-zero sequence number, since 0 is the default value and what we'd expect to get at the end if a Datagram lost the sequence number somewhere in the roundtrip path.